shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/TemporaryTerminalFilter.kt

```diff
@@ -0,0 +1,136 @@
+    // cancellations
+    private val predictionsByRoute =
+        predictions.predictions.values
+            .filterNot { it.scheduleRelationship == Prediction.ScheduleRelationship.Cancelled }
+            .groupBy { it.routeId }
+    private val routePatternsByRoute = globalData.routePatterns.values.groupBy { it.routeId }
+
+    fun Schedule.routePatternId() = schedules.trips[tripId]?.routePatternId
+
+    fun Prediction.routePatternId() = predictions.trips[tripId]?.routePatternId
+
+    /**
+     * Rewriting realtime data for temporary terminals causes what the app displays to diverge from
+     * what the API says is real, and so must be done with caution. Specifically, we only want to
+     * rewrite
+     * 1. subway routes
+     * 2. with alerts somewhere on the line
+     * 3. where the schedule has non-typical patterns and is missing a typical pattern (requiring
+     *    only non-typical patterns misses single-branch RL disruptions, but allowing just any
+     *    non-typical pattern hits GL early-morning cross-branch trips)
+     * 4. but the predictions all have typical patterns instead of non-typical patterns (meaning
+     *    there are no predictions that will be lost)
+     */
+    fun appliesToRoute(route: Route): Boolean {
+        val routeId = route.id
+        val isSubway = route.type.isSubway()
+
+        val routeHasAlert =
+            alerts.any { alert ->
+                alert.effect in setOf(Alert.Effect.Suspension, Alert.Effect.Shuttle) &&
+                    alert.anyInformedEntity { it.appliesTo(routeId = routeId) }
+            }
+
+        val routeSchedules = schedulesByRoute[routeId].orEmpty()
+        val scheduledPatterns = routeSchedules.mapNotNullTo(mutableSetOf()) { it.routePatternId() }
+        val routePatterns = routePatternsByRoute[routeId].orEmpty()
+        val scheduleMissingTypical =
+            routePatterns
+                .filter { it.typicality == RoutePattern.Typicality.Typical }
+                .any { it.id !in scheduledPatterns }
+        val scheduleHasNontypical =
+            scheduledPatterns.any {
+                globalData.routePatterns[it]?.typicality != RoutePattern.Typicality.Typical
+            }
+        val scheduleReplacedTypical = scheduleMissingTypical && scheduleHasNontypical
+
+        val routePredictions = predictionsByRoute[routeId].orEmpty()
+        val predictionsAlwaysTypical =
+            routePredictions.isNotEmpty() &&
+                routePredictions.all {
+                    globalData.routePatterns[it.routePatternId()]?.typicality ==
+                        RoutePattern.Typicality.Typical
+                }
+
+        return isSubway && routeHasAlert && scheduleReplacedTypical && predictionsAlwaysTypical
+    }
+
+    /**
+     * Discards scheduled, non-predicted, non-typical headsigns in this
+     * [NearbyStaticData.StopPatterns].
+     */
+    fun filterPatternsAtStop(
```

Comment:
suggestion (non-blocking): Describing the type of filtering in the function name could be clearer, like `filterScheduleOnlyTemporary`

Reply:
e11776a